### ISSUE TITLE
C output streams not redirected to logfile

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#8108]
as said in the title it seems like std_print is not redirected to the logfile, when i run wesnoth without a console attached those outputs are missing in the logfile. If i remember  correctly this behaviour was different on windows, but not fulyl sure.

When a run wesnoth with a console it shows normally. 

- **Version:* 1.17.23+dev
- **Downloaded from:* self biult
- **Build info:**
- **OS:* ubuntu linux.

**Comment by Pentarctagon** [comment]
Where is std_print implemented in wesnoth?

**Comment by gfgtdf** [comment]
> Where is std_print implemented in wesnoth?

its an alias for luas print function, since print in wesnoth prints to the lua console.

**Comment by Pentarctagon** [comment]
Wesnoth redirects output at https://github.com/wesnoth/wesnoth/blob/ff8d4d063f59a0aa54271184ddbc1e4fc608d13b/src/log.cpp#L181C13-L199

No idea how lua's print works, but presumably it's using a different mechanism.

**Comment by gfgtdf** [comment]
Maybe it only redirects the c++ iostreams but not the C streams (the the c function like printf write to) this woudl mean that all aeerors from third partly libraries we use (SDL/cairo/assertionerrors) would be missing from the logfile.

**Comment by CelticMinstrel** [comment]
Maybe replacing the cerr/cout stream buffer breaks the synchronization? What if you add a call to [this](https://en.cppreference.com/w/cpp/io/ios_base/sync_with_stdio) after swapping in the new stream buffer?

**Comment by soliton-** [comment]
Assertion messages are definitely missing from the log file.

**Comment by gfgtdf** [comment]
> Maybe replacing the cerr/cout stream buffer breaks the synchronization? What if you add a call to [this](https://en.cppreference.com/w/cpp/io/ios_base/sync_with_stdio) after swapping in the new stream buffer?

To me it seems more like the c streams are not redirected at all, while that function onyl handles the case that the order of the output is wrong.

>  Assertion messages are definitely missing from the log file.

Very good to know, i think we can definitely call this a bug then, if someone sends us a logfile of an error we surely  want this important information in it.

I'm also quite sure that when i used to play on windows this problem didn't exist, but that was on 1.14 so its totalyl possibel that it now also affects windows.

**Comment by Pentarctagon** [comment]
Writing logs to file was made consistent across all platforms, so it would affect them all.

**Comment by CelticMinstrel** [comment]
> To me it seems more like the c streams are not redirected at all, while that function onyl handles the case that the order of the output is wrong.

By default it's specified that the C++ and C streams should output to the same place. I'm not sure exactly how that works, but it specifies that the low-level output functions are required to have identical effect between the C stream and the synchronized C++ stream. So surely it's worth at least trying that function to see if it fixes the issue?

**Comment by soliton-** [comment]
`std::cout` writes to `stdout` by default. So if you change `stdout` then that also changes what `std::cout` writes to but not the other way around.

Redirecting stdout/stderr instead should be easy but the boost::iostreams::tee stuff we currently do probably not so much.

**Comment by Pentarctagon** [comment]
The old Windows-specific code did redirecting with https://github.com/wesnoth/wesnoth/blob/1.16/src/log_windows.cpp#L382, if that's helpful as an example.

**Comment by CelticMinstrel** [comment]
> i looked up a possible implementation

Aside from not bothering to implement "turn synching back on", that implementation looks like it follows what I said here:

> It wouldn't be particularly surprising to me if the standard implementers interpreted `cerr.rdbuf()` to mean "the default cerr streambuf" rather than "the current cerr streambuf", in which case this wouldn't work.

Desynching them consists of just constructing a new streambuf, which implies that synching is a feature built into the default streambuf and will never work with any substituted streambuf (unless of course that streambuf itself handles it).

**Comment by CelticMinstrel** [comment]
> I tried to look at this, and i couldnt find out how to do "teeing" under windows unfortuntely, if someone has another an idea please let us know.

Doesn't Boost have a file stream that takes a `FILE*` as the input to wrap an existing C file in the C++ stream machinery? I think that might be the key to getting this to work – just create a stream that wraps `stderr` and tee that in the same way it's doing now.

**Comment by Pentarctagon** [comment]
Not that I'm aware of, at least.

**Comment by Pentarctagon** [comment]
Something like `boost::iostreams::stream_buffer<boost::iostreams::file_descriptor_sink> sink(fileno(stderr), boost::iostreams::never_close_handle);` maybe? I found a partial example over on stack overflow, but I have really no idea how any of it works.

**Comment by CelticMinstrel** [comment]
That sounds like it's pointing in vaguely the right direction, at least. File descriptor isn't _quite_ what I was hoping for, though – does that work on Windows? I think file descriptor is the POSIX level, not the standard C level.

**Comment by Pentarctagon** [comment]
I have no way to check, but yeah I'd guess it probably doesn't work on Windows. Unfortunately trying to get google to return useful results is particularly difficult - it doesn't use the `*` in `FILE*` as a search term, tends to have a ton of results for shell redirection which isn't helpful, and seems to assume that std::cout/stdout and std::cerr/stderr are the same thing as soon as I use `C++` in the search.

**Comment by CelticMinstrel** [comment]
Looking over the Boost documentation, it seems that the file descriptor sink is _somehow_ usable on Windows but you need to provide a native Windows file handle to it. So I guess it can be used in a cross-platform solution. I don't understand why there isn't one that literally just takes a `FILE*` though.

**Comment by CelticMinstrel** [comment]
Other than using Boost's file descriptor device, the other option would be to just create one that uses `FILE*` – either a boost device or a std filebuf. I don't think it would be all that complicated, especially if you use the code for the file descriptor one as a reference.

**Comment by gfgtdf** [comment]
> Doesn't Boost have a file stream that takes a FILE* as the input to wrap an existing C file in the C++ stream machinery? I think that might be the key to getting this to work – just create a stream that wraps stderr and tee that in the same way it's doing now.

The main challenge of this issue is to redirect and "tee" the c output streams stderr/stdout. A c++ stream stink that wraps stderr won't help with that at all since it doesn't change stderr.

> Unfortunately trying to get google to return useful results is particularly difficult - it doesn't use the * in FILE* as a search term, tends to have a ton of results for shell redirection which isn't helpful

Yes i had this problem aswell.

**Comment by Pentarctagon** [comment]
It's a shame the wide version of `freopen` is Windows-specific. Otherwise the old Windows-specific logging implementation for this was pretty simple (though also wouldn't support tee). I assume that using `freopen` on a path containing non-ascii characters would cause problems.

**Comment by gfgtdf** [comment]
> It's a shame the wide version of freopen is Windows-specific

Just use the non wide character version on linux, doesn't really sounds like an issue to me.

>  I assume that using freopen on a path containing non-ascii characters would cause problems.

Only on windows (and also only on windows 7 and early windows 10, same as the fontconfig issue )




Alternatively i have found this solution on stackoverflow which seems to support support tee, 
```
    if(dup2(fileno(popen("tee out.txt", "w")), STDOUT_FILENO) < 0) {
        fprintf(stderr, "couldn't redirect output\n");
        return 1;
    }
```
which seems to support tee, but its probably linux  specific, also i don't really know how these functions work tbh.

**Comment by CelticMinstrel** [comment]
Hmm. What about examining this from a different angle? Instead of trying to make the tee somehow handle `stderr`, maybe we could replace `stderr` with a pipe? Then start up a thread that does nothing but read from the pipe and flush it through to `std::cerr`.

I don't know if there's a portable way to make a pipe, though. Since the idea is to replace `stderr`, it definitely _has_ to an OS-native pipe or it won't work. The other tricky thing would be to link it up to `stderr`, and I'm not sure if that's even doable. Also, would it have to be done on `stdout` as well?

**Comment by gfgtdf** [comment]
> Then start up a thread that does nothing but read from the pipe and flush it through to std::cerr.

Despite being rather complicated, i'm also worried that this would not work properly in case of an error, for example an assert() message where the programm is terminated directly afterwards (and probably before the other thread has time to print it to the output).

**Comment by Pentarctagon** [comment]
Is supporting tee required? IIRC it was added to address the debate over when to log to file vs to cerr, but it's making this problem exceptionally difficult to solve.

> Only on windows (and also only on windows 7 and early windows 10, same as the fontconfig issue )

So for 1.17+, both Windows and non-Windows could use the standard `freopen`, given the minimum recommended Windows version is Windows 10 1903+?

**Comment by gfgtdf** [comment]
> So for 1.17+, both Windows and non-Windows could use the standard freopen, given the minimum recommended Windows version is Windows 10 1903+?

I think so, but its probably better to test.

> Is supporting tee required?

Idk, i wasn't there when i was added, i just assumed it was since i thought it was the main reason why that code was changed in the first place.

**Comment by Pentarctagon** [comment]
The code was originally changed to support writing to a log file on macOS and Linux, rather than just Windows like 1.16 does. Then writing to a log file was made the default, since that's what's most useful to players, but that then meant people who wanted the output written to the terminal needed to add the `--no-log-to-file` option when running Wesnoth. So loonycyborg added the use of tee to just have it write to both.

**Comment by Pentarctagon** [comment]
It would also presumably fix #8255
